### ISSUE TITLE
Retry on connection EOF?

### ISSUE THREAD

**Issue by horgh** [elastic/go-elasticsearch#111]
I'm using the 7.x branch and I've found that sometimes my Search call returns an EOF error. I'm thinking the connection went bad during a period of not being used. Would this be a case we could add to the estransport's retry logic to make it so the caller doesn't need to worry about it?

Thank you!

**Comment by karmi** [comment]
Hello — sorry for the late reply.

I can't rule out that something happens either in the retry/connection management logic which would cause a bug, or that something happens in the lower layers (`net/http`). However, I don't have any immediate idea why this would happen — if the connection returns an error, it's put into the list of dead connections.

Could you please provide more information about the exact version you're using (Git commit), and also more information about your environment (number of nodes, etc)?

I'm wondering if the situation could be replicated by launching a couple of servers via `net/http`, and modelling the situation (eg. closing the connection)?

**Comment by horgh** [comment]
I'm using version v7.4.1. I'll have to get back to you about the number of nodes!

To be clear, I don't think there's a bug in this library. I'm thinking the connection timed out or something (it occurs after hours of not using the connection). I was mainly wondering whether there could be an improvement - whether we could retry in an additional condition: When we see an EOF error when using the connection. i.e., adding an additional condition [here](https://github.com/elastic/go-elasticsearch/blob/master/estransport/estransport.go#L258).

I agree it will be put into the list of dead connections, but my program sees the raised error and currently will need to retry on its own, whereas other network type errors will be retried transparently.

**Comment by karmi** [comment]
Thanks for the additional info!

I see what you mean now — I've tried to run a test program against an Elasticsearch cluster which is still starting (and which replies with `EOF` / `Empty reply from server` in such situation), and the error is not retried.

I _think_ it makes sense to add the condition you suggest — it will also help eg. with retrying a request when the client hits a node which is currently restarting.

However, as I'm about to release a `7.5` package very soon, I'd rather not make a last-minute change like this, without first thinking a little bit more about it.

**Comment by horgh** [comment]
Cool, thank you for considering it! :smile:

**Comment by karmi** [comment]
I've added the `io.EOF` error into the relevant condition checking in the attached commit, and backported into `6.x` and `7.x` branches.

This is a nice catch, since in my experiments, an `io.EOF` is returned when eg. an Elasticsearch node is restarting, which is an expected condition. I've verified locally by launching multiple nodes in Docker containers, and restarting one of those containers. The error response is properly retried, and after the timeout, the dead condition resurrected. Thanks again for the report!
